transformer_engine/pytorch/cpu_offload.py

```diff
@@ -0,0 +1,410 @@
+        
+        # during forward, the next_group_to_fetch always points to the min of 
+        # the last commited group, and the last offloaded group
+        self.next_group_to_fetch = min(self.current_group, self.num_offload_group -1)
+
+        super().on_group_commit_forward()
+
+    def bulk_reload_group(self, group_to_reload):
+        assert group_to_reload < self.num_offload_group
+        if group_to_reload == self.num_offload_group - 1:
+            self.h2d_stream.wait_event(self.d2h_final_event)
+        with torch.cuda.stream(self.h2d_stream):
+            # move back tensors
+            for tensor_label in self.tensor_tag_to_state.keys():
+                group_id, _ = tensor_label
+                if group_id == group_to_reload:
+                    state = self.tensor_tag_to_state[tensor_label]
+                    if isinstance(state, tuple):
+                        recovered_tensor = SynchronizedGroupOffloadHandler.reload(state)
+                        self.tensor_tag_to_state[tensor_label] = recovered_tensor
+                    else:
+                        self.tensor_tag_to_state[tensor_label] = state
+
+    def on_group_commit_backward(self):
+        # first decrement the current group.
+        # after last commit in forward, the group will +1; in backward it -1. Finally it should be decremented to 0
+        self.current_group -= 1
+        assert self.current_group >= 0
+
+        # decide the range of group to prefetch
+        should_prefetch_until_group = self.current_group - self.num_prefetch_group
+        if should_prefetch_until_group < 0:
+            should_prefetch_until_group = 0
+        
+        # do prefetch
+        for group_num_to_prefetch in range(self.next_group_to_fetch, should_prefetch_until_group - 1, -1):
+            # record the event in the compute stream, for h2d to wait
+            torch.cuda.current_stream().record_event(self.compute_stream_bwd_start_events[group_num_to_prefetch])
+            
+            # start of h2d should wait for the compute and the d2h
+            self.h2d_stream.wait_event(self.compute_stream_bwd_start_events[group_num_to_prefetch])
+            
+            #recover tensors (copy back from host)
+            self.bulk_reload_group(group_num_to_prefetch)
+            
+            # record an event for the backward of this layer to wait
+            self.h2d_stream.record_event(self.h2d_finish_events[group_num_to_prefetch])
+        
+        self.next_group_to_fetch = min(self.num_offload_group - 1, should_prefetch_until_group - 1) # always is set to -1 at the end of the backward
+        
+        # wait for the current group
+        if self.current_group < self.num_offload_group:
+            torch.cuda.current_stream().wait_event(self.h2d_finish_events[self.current_group])
+
+def get_cpu_offload_context(cpu_offloading, cpu_offloading_num_layers, cpu_offloading_activations, cpu_offloading_weights):
```

Comment:
Please add the docs to this function (as I believe this to be the user-facing function, right?) and add it to the autodoc here: https://github.com/NVIDIA/TransformerEngine/blob/main/docs/api/pytorch.rst

Please also add it to the __init__.py in transformer_engine/pytorch so that it is importable by
```python
from transformer_engine.pytorch import get_cpu_offload_context
```